File: examples/tlplay-python/MainWindow.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright Contributors to the feather-tk project.

import opentimelineio as otio
import ftkPy as ftk
import tlRenderPy as tl

import FileActions
import Menus
import PlaybackActions
import PlaybackBar
import SettingsWidget
import StatusBar
import ToolBars
import ViewActions
import WindowActions

import weakref

class MainWindow(ftk.MainWindow):
    """
    The main window creates the widgets and actions.
    """
    def __init__(self, context, app):
        ftk.MainWindow.__init__(self, context, app, ftk.Size2I(1280, 960))

        # Restore settings.
        settingsToggle = False
        splitter = 0.8
        splitter2 = 0.8
        self._settingsModel = app.getSettingsModel()
        settings = self._settingsModel.getBool("/MainWindow/SettingsVisible")
        if settings[0]:
            settingsToggle = settings[1]
        settings = self._settingsModel.getDouble("/MainWindow/Splitter")
        if settings[0]:
            splitter = settings[1]
        settings = self._settingsModel.getDouble("/MainWindow/Splitter2")
        if settings[0]:
            splitter2 = settings[1]

        # Create observables.
        self.settingsToggle = ftk.ObservableBool(settingsToggle)

        # Create the main widgets.
        self._viewport = tl.ui.Viewport(context)
        self._timelineWidget = tl.ui.TimelineWidget(context, app.getTimeUnitsModel())

        # Create the actions.
        self._fileActions = FileActions.Actions(context, app, self)
        self._playbackActions = PlaybackActions.Actions(context, app)
        self._viewActions = ViewActions.Actions(context, app, self)
        self._windowActions = WindowActions.Actions(context, app, self)

        # Create the menu bar.
        self._menuBar = ftk.MenuBar(context)
        self._menuBar.addMenu("File", Menus.File(context, app, self._fileActions))
        self._menuBar.addMenu("Playback", Menus.Playback(context, app, self._playbackActions))
        self._menuBar.addMenu("View", Menus.View(context, app, self._viewActions))
        self._menuBar.addMenu("Window", Menus.Window(context, app, self._windowActions))
        self.menuBar = self._menuBar

        # Create the tool bars.
        self._fileToolBar = ToolBars.File(context, self._fileActions)
        self._viewToolBar = ToolBars.View(context, self._viewActions)
        self._windowToolBar = ToolBars.Window(context, self._windowActions)
        self._playbackBar = PlaybackBar.Widget(context, app, self._playbackActions)
        self._statusBar = StatusBar.Widget(context, app, self)

        # Create the settings widget.
        self._settingsWidget = SettingsWidget.Widget(context, app)
        self._settingsWidget.setVisible(settingsToggle)

        # Layout widgets.
        self._layout = ftk.VerticalLayout(context)
        self._layout.spacingRole = ftk.SizeRole._None
        self.setWidget(self._layout)
        hLayout = ftk.HorizontalLayout(context, self._layout)
        hLayout.spacingRole = ftk.SizeRole.SpacingSmall
        self._fileToolBar.parent = hLayout
        ftk.Divider(context, ftk.Orientation.Horizontal, hLayout)
        self._viewToolBar.parent = hLayout
        ftk.Divider(context, ftk.Orientation.Horizontal, hLayout)
        self._windowToolBar.parent = hLayout
        ftk.Divider(context, ftk.Orientation.Vertical, self._layout)
        self._splitter = ftk.Splitter(context, ftk.Orientation.Vertical, self._layout)
        self._splitter.split = splitter
        self._splitter2 = ftk.Splitter(context, ftk.Orientation.Horizontal, self._splitter)
        self._splitter2.split = splitter2
        self._viewport.parent = self._splitter2
        self._settingsWidget.parent = self._splitter2
        vLayout = ftk.VerticalLayout(context, self._splitter)
        vLayout.spacingRole = ftk.SizeRole._None
        self._playbackBar.parent = vLayout
        ftk.Divider(context, ftk.Orientation.Vertical, vLayout)
        self._timelineWidget.parent = vLayout
        ftk.Divider(context, ftk.Orientation.Vertical, self._layout)
        self._statusBar.parent = self._layout

        # Create observers.
        selfWeak = weakref.ref(self)
        self.playerObserver = tl.timeline.PlayerObserver(
            app.getDocumentModel().observePlayer(),
            lambda player: selfWeak()._widgetUpdate(player))

    def __del__(self):
    
        # Save settings.
        self._settingsModel.setBool(
            "/MainWindow/SettingsVisible",
            self.settingsToggle.get())
        self._settingsModel.setDouble("/MainWindow/Splitter", self._splitter.split)
        self._settingsModel.setDouble("/MainWindow/Splitter2", self._splitter2.split)

    def getViewport(self):
        """
        Get the viewport.
        """
        return self._viewport

    def setSettingsVisible(self, value):
        """
        Set whether the settings widget is visible.
        """
        if self.settingsToggle.setIfChanged(value):
            self._settingsWidget.setVisible(value)

    def _widgetUpdate(self, player):
        self._viewport.player = player
        self._timelineWidget.player = player

File: lib/tlRender/PlayApp/MainWindow.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright Contributors to the tlRender project.

#include "MainWindow.h"

#include "App.h"
#include "CompareActions.h"
#include "FileActions.h"
#include "FilesModel.h"
#include "MenuBar.h"
#include "PlaybackActions.h"
#include "PlaybackBar.h"
#include "SettingsModel.h"
#include "SettingsWidget.h"
#include "StatusBar.h"
#include "TabBar.h"
#include "ToolBars.h"
#include "ViewActions.h"
#include "WindowActions.h"

#include <ftk/UI/Divider.h>
#include <ftk/UI/IconSystem.h>
#include <ftk/UI/Menu.h>
#include <ftk/UI/ToolBar.h>

namespace tl
{
    namespace play
    {
        void MainWindow::_init(
            const std::shared_ptr<ftk::Context>& context,
            const std::shared_ptr<App>& app)
        {
            ftk::Window::_init(context, app, "tlplay", ftk::Size2I(1280, 960));

            auto iconSystem = context->getSystem<ftk::IconSystem>();
            setIcon(iconSystem->get("tlRender", 1.0));

            _app = app;
            
            // Restore settings.
            bool settingsVisible = false;
            double splitter = 0.8;
            double splitter2 = 0.8;
            auto settingsModel = app->getSettingsModel();
            settingsModel->get("/MainWindow/SettingsVisible", settingsVisible);
            _settingsVisible = ftk::Observable<bool>::create(settingsVisible);
            settingsModel->get("/MainWindow/Splitter", splitter);
            settingsModel->get("/MainWindow/Splitter2", splitter2);
            _settingsModel = settingsModel;

            // Create the viewport.
            _viewport = ui::Viewport::create(context);
            //timeline::BackgroundOptions bgOptions;
            //bgOptions.type = timeline::Background::Gradient;
            //_viewport->setBackgroundOptions(bgOptions);
            //timeline::ForegroundOptions fgOptions;
            //fgOptions.grid.enabled = true;
            //fgOptions.grid.size = 1;
            //fgOptions.grid.labels = timeline::GridLabels::Alphanumeric;
            //fgOptions.grid.lineWidth = 10;
            //fgOptions.outline.enabled = true;
            //_viewport->setForegroundOptions(fgOptions);
            ftk::ImageOptions imageOptions;
            imageOptions.imageFilters.minify = ftk::ImageFilter::Nearest;
            imageOptions.imageFilters.magnify = ftk::ImageFilter::Nearest;
            _viewport->setImageOptions({ imageOptions });
            timeline::DisplayOptions displayOptions;
            displayOptions.imageFilters.minify = ftk::ImageFilter::Nearest;
            displayOptions.imageFilters.magnify = ftk::ImageFilter::Nearest;
            _viewport->setDisplayOptions({ displayOptions });

            // Create the timeline.
            _timelineWidget = ui::TimelineWidget::create(
                context,
                app->getTimeUnitsModel());
            ui::DisplayOptions timelineDisplayOptions;
            timelineDisplayOptions.minimize = false;
            //timelineDisplayOptions.thumbnails = false;
            _timelineWidget->setDisplayOptions(timelineDisplayOptions);
            _timelineWidget->setVStretch(ftk::Stretch::Expanding);

            // Create the actions.
            _fileActions = FileActions::create(context, app);
            _compareActions = CompareActions::create(context, app);
            _playbackActions = PlaybackActions::create(context, app);
            _viewActions = ViewActions::create(context, app, _viewport);
            _windowActions = WindowActions::create(
                context,
                app,
                std::dynamic_pointer_cast<MainWindow>(shared_from_this()));

            // Create the menu bar.
            _menuBar = MenuBar::create(
                context,
                app,
                _fileActions,
                _compareActions,
                _playbackActions,
                _viewActions,
                _windowActions);

            // Create the tool bars.
            auto toolBars = ToolBars::create(
                context,
                _fileActions,
                _compareActions,
                _viewActions,
                _windowActions);
            _playbackBar = PlaybackBar::create(
                context,
                app,
                _playbackActions->getActions());
            _statusBar = StatusBar::create(context, app);

            // Crate the tab bar.
            _tabBar = TabBar::create(context, app);

            // Crate the settings widget.
            _settingsWidget = SettingsWidget::create(context, app);
            _settingsWidget->setVisible(settingsVisible);

            // Layout widgets.
            _layout = ftk::VerticalLayout::create(context, shared_from_this());
            _layout->setSpacingRole(ftk::SizeRole::None);
            _menuBar->setParent(_layout);
            ftk::Divider::create(context, ftk::Orientation::Vertical, _layout);
            toolBars->setParent(_layout);
            ftk::Divider::create(context, ftk::Orientation::Vertical, _layout);
            _splitter = ftk::Splitter::create(context, ftk::Orientation::Vertical, _layout);
            _splitter->setSplit(splitter);
            _splitter2 = ftk::Splitter::create(context, ftk::Orientation::Horizontal, _splitter);
            _splitter2->setSplit(splitter2);
            auto vLayout = ftk::VerticalLayout::create(context, _splitter2);
            vLayout->setSpacingRole(ftk::SizeRole::None);
            _tabBar->setParent(vLayout);
            _viewport->setParent(vLayout);
            _settingsWidget->setParent(_splitter2);
            vLayout = ftk::VerticalLayout::create(context, _splitter);
            vLayout->setSpacingRole(ftk::SizeRole::None);
            _playbackBar->setParent(vLayout);
            ftk::Divider::create(context, ftk::Orientation::Vertical, vLayout);
            _timelineWidget->setParent(vLayout);
            ftk::Divider::create(context, ftk::Orientation::Vertical, vLayout);
            _statusBar->setParent(vLayout);

            // Create observers.
            _playerObserver = ftk::Observer<std::shared_ptr<timeline::Player> >::create(
                app->getFilesModel()->observePlayer(),
                [this](const std::shared_ptr<timeline::Player>& value)
                {
                    _viewport->setPlayer(value);
                    _timelineWidget->setPlayer(value);
                });

            _compareObserver = ftk::Observer<timeline::Compare>::create(
                app->getFilesModel()->observeCompare(),
                [this](timeline::Compare value)
                {
                    timeline::CompareOptions options;
                    options.compare = value;
                    _viewport->setCompareOptions(options);
                });
        }

        MainWindow::~MainWindow()
        {
            // Save settings.
            if (auto settingsModel = _settingsModel.lock())
            {
                settingsModel->set(
                    "/MainWindow/SettingsVisible",
                    _settingsVisible->get());
                settingsModel->set("/MainWindow/Splitter", _splitter->getSplit());
                settingsModel->set("/MainWindow/Splitter2", _splitter2->getSplit());
            }
        }

        std::shared_ptr<MainWindow> MainWindow::create(
            const std::shared_ptr<ftk::Context>& context,
            const std::shared_ptr<App>& app)
        {
            auto out = std::shared_ptr<MainWindow>(new MainWindow);
            out->_init(context, app);
            return out;
        }

        const std::shared_ptr<ui::Viewport>& MainWindow::getViewport() const
        {
            return _viewport;
        }

        std::shared_ptr<ftk::IObservable<bool> > MainWindow::observeSettingsVisible() const
        {
            return _settingsVisible;
        }

        void MainWindow::setSettingsVisible(bool value)
        {
            if (_settingsVisible->setIfChanged(value))
            {
                _settingsWidget->setVisible(value);
            }
        }

        void MainWindow::keyPressEvent(ftk::KeyEvent& event)
        {
            event.accept = _menuBar->shortcut(event.key, event.modifiers);
        }

        void MainWindow::keyReleaseEvent(ftk::KeyEvent& event)
        {
            event.accept = true;
        }

        void MainWindow::dropEvent(ftk::DragDropEvent& event)
        {
            event.accept = true;
            if (auto textData = std::dynamic_pointer_cast<ftk::DragDropTextData>(event.data))
            {
                if (auto app = _app.lock())
                {
                    for (const auto& fileName : textData->getText())
                    {
                        app->open(ftk::Path(fileName));
                    }
                }
            }
        }
    }
}
